sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -190,9 +190,13 @@ object FileFormatWriter extends Logging {
           global = false,
           child = plan).execute()
       }
-      val ret = new Array[WriteTaskResult](rdd.partitions.length)
+
+      // SPARK-23271 If we are attempting to write a zero partition rdd, change the number of
+      // partition to 1 to make sure we at least set up one write task to write the metadata.
+      val finalRdd = if (rdd.partitions.length == 0) rdd.repartition(1) else rdd
```

Comment:
Looks like shuffle will be here if partitions number is zero. If so, maybe, other solution is possible?

Reply:
Yea the shuffle can be avoided. We can just launch a write task for empty RDD, instead of calling `rdd.repartition(1)`.

Reply:
@cloud-fan @pashazm I was thinking, this would not be a regular event to write empty datasets , right ? Should we be even optimizing this path ? Secondly, is shuffling an empty data set that expensive ?

@cloud-fan, actually i had tried to launch a write task for empty RDD, but was hitting a NullPointerException from scheduler ? Looks like things are setup to only work off of partitions of RDD. Could we try to create this empty metadata file from the driver in this case ? If we go that route, then we may have to refactor the write task code. Seems like a lot for this little corner case, what do you think ?

Reply:
You could try `coalesce(1)` that should not shuffle.

Reply:
@hvanhovell Thanks. I have a question. Can we go from zero partition to one partition with coalesce() ? In the code we seem to be doing a min(prevPartition, requestedPartition) to set the target number of partition [code](
https://github.com/apache/spark/blob/e26dac5feb02033f980b1e69c9b0ff50869b6f9e/core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala#L393 )

Reply:
@hvanhovell Just tried. We stay at numPartitions = 0 after coalesce(). So it does not fix the problem.

Reply:
One simple way to fix it: create an empty 1-partition RDD and use it here.

Reply:
Yea, you can have
```
sparkSession.sparkContext.parallelize(Array.empty[InternalRow])
```

Reply:
@cloud-fan @jiangxb1987 Thanks a **LOT**. This works perfectly.